File: front/packages/models/src/resources/movie.ts
/*
 * Kyoo - A portable and vast media library solution.
 * Copyright (c) Kyoo.
 *
 * See AUTHORS.md and LICENSE file in the project root for full license information.
 *
 * Kyoo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * Kyoo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Kyoo. If not, see <https://www.gnu.org/licenses/>.
 */

import { z } from "zod";
import { ImagesP, ResourceP, imageFn } from "../traits";
import { zdate } from "../utils";
import { CollectionP } from "./collection";
import { Genre } from "./genre";
import { MetadataP } from "./metadata";
import { Status } from "./show";
import { StudioP } from "./studio";
import { WatchStatusP } from "./watch-status";

export const MovieP = ResourceP("movie")
	.merge(ImagesP)
	.extend({
		/**
		 * The title of this movie.
		 */
		name: z.string(),
		/**
		 * A catchphrase for this show.
		 */
		tagline: z.string().nullable(),
		/**
		 * The list of alternative titles of this movie.
		 */
		aliases: z.array(z.string()),
		/**
		 * The summary of this movie.
		 */
		overview: z.string().nullable(),
		/**
		 * A list of tags that match this movie.
		 */
		tags: z.array(z.string()),
		/**
		 * /** Is this movie not aired yet or finished?
		 */
		status: z.nativeEnum(Status),
		/**
		 * How well this item is rated? (from 0 to 100).
		 */
		rating: z.number().int().gte(0).lte(100),
		/**
		 * How long is this movie? (in minutes).
		 */
		runtime: z.number().int().nullable(),
		/**
		 * The date this movie aired. It can also be null if this is unknown.
		 */
		airDate: zdate().nullable(),
		/**
		 * A youtube url for the trailer.
		 */
		trailer: z.string().optional().nullable(),
		/**
		 * The list of genres (themes) this movie has.
		 */
		genres: z.array(z.nativeEnum(Genre)),
		/**
		 * The studio that made this movie.
		 */
		studio: StudioP.optional().nullable(),
		/**
		 * The collection this movie is part of.
		 */
		collections: z.array(CollectionP).optional(),
		/**
		 * The links to see a movie or an episode.
		 */
		links: z.object({
			/**
			 * The direct link to the unprocessed video (pristine quality).
			 */
			direct: z.string().transform(imageFn),

			/**
			 * The link to an HLS master playlist containing all qualities available for this video.
			 */
			hls: z.string().transform(imageFn).nullable(),
		}),
		/**
		 * The link to metadata providers that this show has.
		 */
		externalId: MetadataP,
		/**
		 * Metadata of what an user as started/planned to watch.
		 */
		watchStatus: WatchStatusP.optional().nullable(),
	})
	.transform((x) => ({
		...x,
		runtime: x.runtime === 0 ? null : x.runtime,
	}))
	.transform((x) => {
		if (!x.thumbnail && x.poster) {
			x.thumbnail = { ...x.poster };
			if (x.thumbnail) {
				x.thumbnail.low = x.thumbnail.high;
				x.thumbnail.medium = x.thumbnail.high;
			}
		}
		return x;
	})
	.transform((x) => ({
		...x,
		href: `/movie/${x.slug}`,
		playHref: `/movie/${x.slug}/watch`,
	}));

/**
 * A Movie type
 */
export type Movie = z.infer<typeof MovieP>;

File: back/src/Kyoo.Abstractions/Models/Resources/Movie.cs
// Kyoo - A portable and vast media library solution.
// Copyright (c) Kyoo.
//
// See AUTHORS.md and LICENSE file in the project root for full license information.
//
// Kyoo is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// any later version.
//
// Kyoo is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Kyoo. If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using Kyoo.Abstractions.Controllers;
using Kyoo.Abstractions.Models.Attributes;
using Kyoo.Utils;
using Newtonsoft.Json;

namespace Kyoo.Abstractions.Models
{
	/// <summary>
	/// A series or a movie.
	/// </summary>
	public class Movie : IQuery, IResource, IMetadata, IOnMerge, IThumbnails, IAddedDate, ILibraryItem, INews
	{
		public static Sort DefaultSort => new Sort<Movie>.By(x => x.Name);

		/// <inheritdoc />
		public Guid Id { get; set; }

		/// <inheritdoc />
		[MaxLength(256)]
		public string Slug { get; set; }

		/// <summary>
		/// The title of this show.
		/// </summary>
		public string Name { get; set; }

		/// <summary>
		/// A catchphrase for this movie.
		/// </summary>
		public string? Tagline { get; set; }

		/// <summary>
		/// The list of alternative titles of this show.
		/// </summary>
		public string[] Aliases { get; set; } = Array.Empty<string>();

		/// <summary>
		/// The path of the movie video file.
		/// </summary>
		public string Path { get; set; }

		/// <summary>
		/// The summary of this show.
		/// </summary>
		public string? Overview { get; set; }

		/// <summary>
		/// A list of tags that match this movie.
		/// </summary>
		public string[] Tags { get; set; } = Array.Empty<string>();

		/// <summary>
		/// The list of genres (themes) this show has.
		/// </summary>
		public Genre[] Genres { get; set; } = Array.Empty<Genre>();

		/// <summary>
		/// Is this show airing, not aired yet or finished?
		/// </summary>
		public Status Status { get; set; }

		/// <summary>
		/// How well this item is rated? (from 0 to 100).
		/// </summary>
		public int Rating { get; set; }

		/// <summary>
		/// How long is this movie? (in minutes)
		/// </summary>
		public int Runtime { get; set; }

		/// <summary>
		/// The date this movie aired.
		/// </summary>
		public DateTime? AirDate { get; set; }

		/// <inheritdoc />
		public DateTime AddedDate { get; set; }

		/// <inheritdoc />
		public Image? Poster { get; set; }

		/// <inheritdoc />
		public Image? Thumbnail { get; set; }

		/// <inheritdoc />
		public Image? Logo { get; set; }

		/// <summary>
		/// A video of a few minutes that tease the content.
		/// </summary>
		public string? Trailer { get; set; }

		/// <inheritdoc />
		public Dictionary<string, MetadataId> ExternalId { get; set; } = new();

		/// <summary>
		/// The ID of the Studio that made this show.
		/// </summary>
		[SerializeIgnore] public Guid? StudioId { get; set; }

		/// <summary>
		/// The Studio that made this show.
		/// </summary>
		[LoadableRelation(nameof(StudioId))] public Studio? Studio { get; set; }

		/// <summary>
		/// The list of people that made this show.
		/// </summary>
		[SerializeIgnore] public ICollection<PeopleRole>? People { get; set; }

		/// <summary>
		/// The list of collections that contains this show.
		/// </summary>
		[SerializeIgnore] public ICollection<Collection>? Collections { get; set; }

		/// <summary>
		/// Links to watch this movie.
		/// </summary>
		public VideoLinks Links => new()
		{
			Direct = $"/video/movie/{Slug}/direct",
			Hls = $"/video/movie/{Slug}/master.m3u8",
		};

		/// <inheritdoc />
		public void OnMerge(object merged)
		{
			if (People != null)
			{
				foreach (PeopleRole link in People)
					link.Movie = this;
			}
		}

		public Movie() { }

		[JsonConstructor]
		public Movie(string name)
		{
			if (name != null)
			{
				Slug = Utility.ToSlug(name);
				Name = name;
			}
		}
	}
}
